### Resources 



[Learning BAM file](https://davetang.github.io/learning_bam_file/)
 
 
[SAM file format](https://davetang.org/wiki/tiki-index.php?page=SAM)

[SAM file specifications](http://samtools.github.io/hts-specs/SAMv1.pdf)


##### SAM file 

The Sequence Alignment/Map (SAM) format is a generic nucleotide alignment format that describes the alignment of query sequences or sequencing reads to a reference sequence or assembly. Importantly:

It is flexible enough to store all the alignment information generated by various alignment programs;
It is simple enough to be easily generated by alignment programs or converted from existing alignment formats;
It is compact in file size;
It allows most of the operations on the alignment to work on a stream without loading the whole alignment into memory;
It allows the file to be indexed by genomic position to efficiently retrieve all reads aligning to a locus.

SAM is a tab-delimited text format. SAM is a bit slow to parse; so there is a binary equivalent to SAM, called BAM

samtools view NA06984.chrom16.ILLUMINA.bwa.CEU.low_coverage.20100517.bam | head -1

SRR035022.2621862 163 16 59999 37 22S54M = 60102 179 CCAACCCAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCGACCCTCACCCTCACCC >AAA=>?AA>@@B@B?AABAB?AABAB?AAC@B?@AB@A?A>A@A?AAAAB??ABAB?79A?AAB;B?@?@<=8:8 XT:A:M XN:i:2 SM:i:37 AM:i:37 XM:i:0 XO:i:0 XG:i:0 RG:Z:SRR035022 NM:i:2 MD:Z:0N0N52 OQ:Z:CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCBCCCCCCBBCC@CCCCCCCCCCACCCCC;CCCBBC?CCCACCACA@

 
 
 In the SAM format, each alignment line typically represents the linear alignment of a segment. Each line consists of 11 or more TAB-separated fields. The first eleven fields are always present and in the order shown below; if the information represented by any of these fields is unavailable, that field’s value will be a placeholder, either ‘0’ or ‘*’ as determined by the field’s type. 
 sequence_string.sam
<QNAME> <FLAG> <RNAME> <POS> <MAPQ> <CIGAR> <MRNM> <MPOS> <ISIZE> <SEQ> <QUAL> [<TAG>:<VTYPE>:<VALUE> [...]]

 
 All optional fields follow the TAG:TYPE:VALUE format where TAG is a two-character string that matches
/[A-Za-z][A-Za-z0-9]/. Within each alignment line, no TAG may appear more than once and the order
in which the optional fields appear is not significant.
 
 
 consensus_read_ATCTCGTTT_0_9990_0_10120_1	99	chr1	10001	8	10S113M8S	=	10003	113	TGAAGACGCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCAACCCCTCTGGAT	^^^^^^^^^^^^^^^^^^^^^^^F^^^^^^^^^^^^^^^^^^^^^^^S^S^^^^SS^^S^^^^^S^^^^^^^^^^^SS^^^^S^^F^^S^^^^^F^^^^^^^^S^^^F^S^^FF^S^^^^S^^^^S;S^^S	RG:Z:20220930_sample	AS:i:113	XS:i:100	NM:i:0	sd:f:0	RX:Z:ATCTCGTTT	XV:i:2	XW:i:0

 
![image](https://user-images.githubusercontent.com/20830666/209197886-cb871e93-9923-4adf-a1b1-572ec8cc2ec5.png)

For Optional Field, can check [this](https://samtools.github.io/hts-specs/SAMtags.pdf)
 
Note: Illumina collaped-BAM files ( Read collapsing adds the following BAM tags):
 
 
• RX/XU—UMI.
 
• XV—Number of reads in the family.
 
• XW—Number of reads in the duplex-family, or 0 if not a duplex family.

 
##### Understand BAM flags 

 
 [We can use this tool to convert the flag](http://broadinstitute.github.io/picard/explain-flags.html)
 
SAMtools provides various (sub)tools for manipulating alignments in the SAM/BAM format. The SAM (Sequence Alignment/Map) format (BAM is just the binary form of SAM) is currently the de facto standard for storing large nucleotide sequence alignments

 
 Samtools [please refer to the release notes.](https://github.com/samtools/samtools/releases)
 
 
### Useful commands 

1. samtools --help 
 
 
 Commands:
  -- Indexing
     dict           create a sequence dictionary file
     faidx          index/extract FASTA
     fqidx          index/extract FASTQ
     index          index alignment

  -- Editing
     calmd          recalculate MD/NM tags and '=' bases
     fixmate        fix mate information
     reheader       replace BAM header
     targetcut      cut fosmid regions (for fosmid pool only)
     addreplacerg   adds or replaces RG tags
     markdup        mark duplicates

  -- File operations
     collate        shuffle and group alignments by name
     cat            concatenate BAMs
     merge          merge sorted alignments
     mpileup        multi-way pileup
     sort           sort alignment file
     split          splits a file by read group
     quickcheck     quickly check if SAM/BAM/CRAM file appears intact
     fastq          converts a BAM to a FASTQ
     fasta          converts a BAM to a FASTA

  -- Statistics
     bedcov         read depth per BED region
     depth          compute the depth
     flagstat       simple stats
     idxstats       BAM index stats
     phase          phase heterozygotes
     stats          generate stats (former bamcheck)

  -- Viewing
     flags          explain BAM flags
     tview          text alignment viewer
     view           SAM<->BAM<->CRAM conversion
     depad          convert padded BAM to unpadded BAM

 
 
 2. Convert SAM directly to sorted BAM : 
 
 ```
 samtools index test_sorted.bam test_sorted.bai
 ```
 
 
 3. Filtering out unmapped reads in BAM files
```
 samtools view -h -F 4 blah.bam > blah_only_mapped.sam
 ```
 
 4. Extract BAM entries mapping to specific loci
 
```

 samtools view test.bam chr1:200000-500000
 ```
 
#all reads mapping on chr1 as another bam
 
 ```
samtools view -b test.bam chr1 > test_chr1.bam
 ```


