### Resources 



[Learning BAM file](https://davetang.github.io/learning_bam_file/)
 
 
[SAM file format](https://davetang.org/wiki/tiki-index.php?page=SAM)

[SAM file specifications](http://samtools.github.io/hts-specs/SAMv1.pdf)


SAM 

The Sequence Alignment/Map (SAM) format is a generic nucleotide alignment format that describes the alignment of query sequences or sequencing reads to a reference sequence or assembly. Importantly:

It is flexible enough to store all the alignment information generated by various alignment programs;
It is simple enough to be easily generated by alignment programs or converted from existing alignment formats;
It is compact in file size;
It allows most of the operations on the alignment to work on a stream without loading the whole alignment into memory;
It allows the file to be indexed by genomic position to efficiently retrieve all reads aligning to a locus.

SAM is a tab-delimited text format. SAM is a bit slow to parse; so there is a binary equivalent to SAM, called BAM

samtools view NA06984.chrom16.ILLUMINA.bwa.CEU.low_coverage.20100517.bam | head -1

SRR035022.2621862 163 16 59999 37 22S54M = 60102 179 CCAACCCAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCGACCCTCACCCTCACCC >AAA=>?AA>@@B@B?AABAB?AABAB?AAC@B?@AB@A?A>A@A?AAAAB??ABAB?79A?AAB;B?@?@<=8:8 XT:A:M XN:i:2 SM:i:37 AM:i:37 XM:i:0 XO:i:0 XG:i:0 RG:Z:SRR035022 NM:i:2 MD:Z:0N0N52 OQ:Z:CCCCCCCCCCCCCCCCCCCCCCCCCCCCCCBCCCCCCBBCC@CCCCCCCCCCACCCCC;CCCBBC?CCCACCACA@

The alignment section consists of multiple TAB-delimited lines with each line describing an alignment. Each line is:

sequence_string.sam
<QNAME> <FLAG> <RNAME> <POS> <MAPQ> <CIGAR> <MRNM> <MPOS> <ISIZE> <SEQ> <QUAL> [<TAG>:<VTYPE>:<VALUE> [...]]
 

 
Understand BAM flags 

 
 [We can use this tool to convert the flag](http://broadinstitute.github.io/picard/explain-flags.html)
 
SAMtools provides various (sub)tools for manipulating alignments in the SAM/BAM format. The SAM (Sequence Alignment/Map) format (BAM is just the binary form of SAM) is currently the de facto standard for storing large nucleotide sequence alignments

 
 Samtools [please refer to the release notes.](https://github.com/samtools/samtools/releases)
 
 
### Useful commands 

1. 
